File: python/p0242_valid_anagram/solutions/two_hashmap_frequency_count.py
# Time complexity: O(m + n), O(n) simplified
# Space complexity: O(m + n), O(n) simplified
# where m = len(s), n = len(t)


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:

        if len(s) != len(t):
            return False

        s_freq_map, t_freq_map = {}, {}

        for i in range(len(s)):
            s_freq_map[s[i]] = s_freq_map.get(s[i], 0) + 1
            t_freq_map[t[i]] = t_freq_map.get(t[i], 0) + 1

        return s_freq_map == t_freq_map

File: csharp/P0242_ValidAnagram/Solutions/TwoHashmapFrequencyCount.cs
// Time complexity: O(n + k log k), practically O(n) if k is small
// Space complexity: O(k)
// where k = number of unique characters

namespace P0242_ValidAnagram.Solutions.TwoHashmapFrequencyCount;

public class Solution
{
    public bool IsAnagram(string s, string t)
    {
        if (s.Length != t.Length)
            return false;

        Dictionary<char, int> sFreqMap = new Dictionary<char, int>();
        Dictionary<char, int> tFreqMap = new Dictionary<char, int>();

        for (int i = 0; i < s.Length; i++)
        {
            sFreqMap[s[i]] = sFreqMap.GetValueOrDefault(s[i], 0) + 1;
            tFreqMap[t[i]] = tFreqMap.GetValueOrDefault(t[i], 0) + 1;
        }

        return sFreqMap.Count == tFreqMap.Count && sFreqMap.OrderBy(kv => kv.Key).SequenceEqual(tFreqMap.OrderBy(kv => kv.Key));
    }
}